spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -0,0 +1,334 @@
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.MessageHandler;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.SubscribableChannel;
+import org.springframework.util.Assert;
+
+/**
+ * {@link AbstractBinder} that serves as base class for {@link MessageChannel}
+ * binders. Implementors must implement the following methods:
+ * <ul>
+ * <li>{@link #createProducerDestinationIfNecessary(String, ProducerProperties)}</li>
+ * <li>{@link #createProducerMessageHandler(String, ProducerProperties)} </li>
+ * <li>{@link #createConsumerDestinationIfNecessary(String, String, ConsumerProperties)} </li>
+ * <li>{@link #createConsumerEndpoint(String, String, Object, ConsumerProperties)}</li>
+ * </ul>
+ * @author Marius Bogoevici
+ * @since 1.1
+ */
+public abstract class AbstractMessageChannelBinder<C extends ConsumerProperties, P extends ProducerProperties>
+		extends AbstractBinder<MessageChannel, C, P> {
+
+	protected static final ExpressionParser EXPRESSION_PARSER = new SpelExpressionParser();
+
+	private final EmbeddedHeadersMessageConverter embeddedHeadersMessageConverter = new
+			EmbeddedHeadersMessageConverter();
+
+	/**
+	 * Indicates whether the implementation and the message broker have
+	 * native support for message headers. If false, headers will be
+	 * embedded in the message payloads.
+	 */
+	private final boolean supportsHeadersNatively;
+
+	/**
+	 * Indicates what headers are to be embedded in the payload if
+	 * {@link #supportsHeadersNatively} is true.
+	 */
+	private final String[] headersToEmbed;
+
+	public AbstractMessageChannelBinder(boolean supportsHeadersNatively, String[] headersToEmbed) {
+		this.supportsHeadersNatively = supportsHeadersNatively;
+		this.headersToEmbed = headersToEmbed;
+	}
+
+	/**
+	 * Binds an outbound channel to a given destination. The implementation delegates to
+	 * {@link #createProducerDestinationIfNecessary(String, ProducerProperties)}
+	 * and {@link #createProducerMessageHandler(String, ProducerProperties)} for
+	 * handling the middleware specific logic.
+	 * @param destination        the name of the destination
+	 * @param outputChannel      the channel to be bound
+	 * @param producerProperties the {@link ProducerProperties} of the binding
+	 * @return the Binding for the channel
+	 */
+	@Override
+	public final Binding<MessageChannel> doBindProducer(final String destination, MessageChannel outputChannel,
+			final P producerProperties) {
+		Assert.isInstanceOf(SubscribableChannel.class, outputChannel,
+				"Binding is supported only for SubscribableChannel instances");
+		createProducerDestinationIfNecessary(destination, producerProperties);
+		final MessageHandler producerMessageHandler;
+		try {
+			producerMessageHandler = createProducerMessageHandler(destination, producerProperties);
+		}
+		catch (Exception e) {
```

Comment:
should add a `throws BinderException` clause to the method definition for clarity (and corresponding `@throws` in JavaDoc)